## panel 3 uses XREF files from ensembl to pool all IDs mapping to 3953 stable IDs. Below, the UniProtKB idat mapping file is used to pool all ENSP IDs (stable|isoform UKB to stable ENSP mapping) and filter 5 ensembl files
5.23.2020 @mfpfox
1. searched for stable ID match to ENSP, **1489** stable IDs found with xref ENSP
2. used canonical # and isoform # from panel 2 analysis to search for canonical isoform ID matches to ENSP xref
3. for all IDs found in 01_2018 idat mapping file, used set of ENSP IDs to filter RELEASE multimapping and distance from UKB seq. files

#### QUESTIONS:
- why are some stable IDs in idat found for ENSP xref and others are not, does it have to do with single/multi-isoform UKB status?
    ex. found = 
    ex. not found = 
- how many ENSP were identical to UKB seq. in all releases? 

## results
- 3953 UKB ID = 1466 single isoform and 2487 multi isoform IDs
- i only have caniso IDs for 2423 out of 2487 multi isoform UKB IDs in 3953 total [fig 2 data]
- searching 3953 stable IDs for ENSP xref in idat, i found 1461 single iso UKB IDs and 28 multi iso UKB IDs, total 1489 with stable ID xref in idat. this is 62% of total 3953
- for remaining 2464 UKB IDs with no stable ID match found in idat (5 single iso IDs and 2459 multi iso IDs), i mapped multi iso IDs to canonical isoform ID number and re searched idat
    - 2408 multi iso IDs found in re-search (remember 28 multi iso found in stable ID search already) of idat with canisoID, this # + 1489 with stable ID xref totals 3897 UKB ID with xref ENSP from idat, or 98.58% out of 3953 total
    - difference of 56 IDs- 51 of these are multi-iso ID with no canisoID (26/51 have no known caniso and 25/51 have canisoID but no ENSP xref in idat) & remaining 5 are single iso IDs with no stable ID xref found in idat
    - total 1489 (stable ID match in idat) + 2408 (caniso ID match in idat) = 3897

**Requiring all multi-isoform UKBIDs to have ENSP xref using canonical isoform ID**
- single iso IDs still 1461
- multi iso IDs with xref from canisoID found in idat = 2418
- 1461 + 2418 = 3879
    - this is 18 IDs less than 3897 IDs found before requirement of canisoID xref in idat for multi isoform associated UKB IDs
    


In [2]:
# packages
import os
import sys
import numpy as np
import pandas as pd
sys.path.append("/Users/mariapalafox/Desktop/Toolbox/")
from all_funx import *
from IPython.core.display import display, HTML
display(HTML("<style>.container {width:90% !important;}</style>"))
from IPython.display import display
pd.set_option('display.max_columns', None)
pd.options.display.max_seq_items = 2000
# set path
os.chdir("/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/HUMAN_9606_uniprot_IDmapping/2018_08_01_IDmapping_download/idat_xref_for_ENSP_IDs/")
print(os.listdir())

['idat_stableID_xref_NOTFOUND.csv', 'UKBIDs_3897_idat_xrefENSP_dropped56_multiisoform_UKBIDs.csv', 'UKBIDs_56dropped_no_xrefENSP_from_idatfile_multiisoformUKBIDs.csv', 'UKBIDs_no_idat_xref_no_canisoID_56multi_isoform_sUKBIDs.csv', 'MULTI_MAPPING_UKBID_xref_ENSP_from_idatmapping_3879UKBIDs_6069ENSP.csv', 'idat_ENSP_filtered_comborelease_FALSEseq_identity_253row_66UKBIDs_95ENSPIDs.csv', 'idat_stableID_xref_FOUND.csv', 'idat_ENSP_filtered_comborelease_searched6069ENSP_for3879UKBID.csv', 'UKB_ISOFORM_GROUP_IDkey.csv', 'MULTI_MAPPING_UKBID_xref_ENSP_from_idatmapping_3879UKBIDs_6069ENSP copy.csv', 'merge_idat_w_multi_isoform_ukbcanonicalID_3524ENSP_2423UKBID.csv', 'idat_merge_3953.csv', 'UKBIDs_3953_xrefENSP_from_idatfile_6241rows.csv', 'combo_allreleases_3953UKBIDs_diffENSPcount_52417rows_isoformgroups.csv', 'VERSION_MAPPING_UKBID_xref_ENSP_from_idatmapping_3879UKBIDs_6069ENSP.csv', 'UKBIDs_yes_idat_sUKBfound_xref_no_canisoID_26multi_isoform_sUKBIDs.csv']


In [12]:
#isoform_labels = pd.read_csv("UKB_ISOFORM_GROUP_IDkey.csv")
print(isoform_labels.shape)
# drop redundancy in file
iso_dropped = isoform_labels.drop_duplicates(keep='first',inplace=False)
iso_dropped.to_csv("UKB_ISOFORM_GROUP_IDkey.csv", index=False)
print(iso_dropped.shape)
#idat_stable_found = pd.read_csv("idat_stableID_xref_FOUND.csv")
#idat_stable_NOTfound = pd.read_csv("idat_stableID_xref_NOTFOUND.csv")
#iso_dropped.reset_index(inplace=True, drop=True)
iso_dropped.columns = ['sUKBID', 'isoform_group']
iso_dropped.head()

(52417, 2)
(3953, 2)


In [200]:
idat_stable_NOTfound.columns = ['sUKBID', 'sUKBID_xref_in_idat']
idat_stable_NOTfound.head(2)
describeMe(idat_stable_found)

(2616, 3)
Index(['sUKBID', 'xref_ID', 'sUKBID_xref_in_idat'], dtype='object')
sUKBID                 0
xref_ID                0
sUKBID_xref_in_idat    0
dtype: int64


In [20]:
print(idat_stable_found.shape)
print(idat_stable_NOTfound.shape)
idat2_stable_found = pd.merge(idat_stable_found, iso_dropped, how='inner', on=['sUKBID'])
print(idat2_stable_found.shape)
idat2_stable_NOTfound = pd.merge(idat_stable_NOTfound, iso_dropped, how='inner', on=['sUKBID'])
print(idat2_stable_NOTfound.shape)

(2616, 3)
(2464, 2)
(2616, 4)
(2464, 3)


## answering question: how many stable IDs found in idat file with a xref for ENSP ID are single or multi-isoform group UKB IDs...

In [33]:
# simplify df then drop redundant rows before count single and multi labels
count_idatfound = idat2_stable_found[['sUKBID', 'isoform_group']].copy()
count_idatfound.drop_duplicates(inplace=True, keep='first')
checkColumnValues(count_idatfound, 'isoform_group')
print("total unique UKB IDs in idat found set : ", 1461+28)
# STABLE IDS FOUND IN IDAT WITH ENSP XREF:

  isoform_group  Count
0        single   1461
1         multi     28
total unique UKB IDs in idat found set :  1489


In [34]:
# simplify df then drop redundant rows before count single and multi labels
count_idatnot = idat2_stable_NOTfound[['sUKBID', 'isoform_group']].copy()
count_idatnot.drop_duplicates(inplace=True, keep='first')
checkColumnValues(count_idatnot, 'isoform_group')
print("total unique UKB IDs in idat NOT FOUND set : ", 2459+5)
# STABLE IDS NOT FOUND IN IDAT WITH ENSP XREF

  isoform_group  Count
0         multi   2459
1        single      5
total unique UKB IDs in idat NOT FOUND set :  2464


---

* 1489 UKB IDs have a xref ENSP from stable ID match in idat
    * 1461 single isoform and 28 multi isoform
* 2464 UKB IDs dont have an xref for stable ID, and below i search idat file but instead use the canonical isoform ID
 
* total ID sum is 3953

---

variables **idat2_stable_found** and **idat2_stable_NOTfound**
- next adding ENSP xref obtained by adding canonical isoform ID info (from panel 2 analysis) along with isoform count and searching for **canonical isoform ID** matches in idat file

In [31]:
canisoID = pd.read_csv("merge_idat_w_multi_isoform_ukbcanonicalID_3524ENSP_2423UKBID.csv")
#print(canisoID.columns)
canisoID = canisoID[['isoformID', 'xref_ID', 'ID', 'fasta_len','canonical_iso_number']]
canisoID.columns = ['canisoID', 'xref_ID', 'sUKBID', 'isoform_count', 'canonical_iso_number']
canisoID.head(2)
sim_caniso = canisoID[['sUKBID','canisoID','xref_ID']]
sim_caniso.head(2)

sUKBID  canisoID          xref_ID
0  P31946  P31946-1  ENSP00000300161
1  P31946  P31946-1  ENSP00000361930

In [32]:
print("stable ID found in idat file: ",idat2_stable_found.shape)
print("stable ID not found in idat file shape: ", idat2_stable_NOTfound.shape)
print("following xref merge, the shape of idat files should not change!")

stable ID found in idat file:  (2616, 4)
stable ID not found in idat file shape:  (2464, 3)
following xref merge, the shape of idat files should not change!


In [36]:
print(len(sim_caniso.sUKBID.unique()))

2423


In [39]:
idat_canisoID = pd.merge(idat2_stable_NOTfound, sim_caniso, on=['sUKBID'], how='left')
idat_canisoID

sUKBID  sUKBID_xref_in_idat isoform_group  canisoID          xref_ID
0     A0AVT1                False         multi  A0AVT1-1  ENSP00000313454
1     A0FGR8                False         multi       NaN              NaN
2     A0JNW5                False         multi  A0JNW5-1  ENSP00000279907
3     A0MZ66                False         multi  A0MZ66-1  ENSP00000347532
4     A1KXE4                False         multi  A1KXE4-1  ENSP00000374565
...      ...                  ...           ...       ...              ...
3518  Q9Y6K9                False         multi  Q9Y6K9-1  ENSP00000471166
3519  Q9Y6K9                False         multi  Q9Y6K9-1  ENSP00000479662
3520  Q9Y6W5                False         multi  Q9Y6W5-1  ENSP00000483313
3521  Q9Y6W6                False         multi  Q9Y6W6-1  ENSP00000355866
3522  Q9Y6Y8                False         multi  Q9Y6Y8-1  ENSP00000358071

[3523 rows x 5 columns]

---
---
---

select rows where canisoID is NAN and filter for non-redundant list of UKB IDs

In [33]:
# funx in maplib.py
def map_column_values(refdf, df, dickey, dicvalue):
    # make sure refdf has dickey and dicvalue column
    # df must have dickey column 'ID'
    newcol = dicvalue # 'HGNC_name'
    refdic = dict(zip(refdf[dickey], refdf[dicvalue])) # makes dict
    df[newcol] = df[dickey] # new col 'gene' name with df key values
    df[dicvalue] = df[dicvalue].map(refdic)
    return df

def add_genename(df, colkey):
    genefile = "/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/TSV_UNIPROT_xref/IDs/GENE_ID_KEY_NOEXCEL_OPENING.csv"
    genekey=pd.read_csv(genefile, usecols=[0,3]) # 0=ID 3=HGNC_name
    df2 = map_column_values(genekey, df, 'sUKBID', 'HGNC_name')
    return df2


In [176]:
noxref = idat_canisoID[idat_canisoID['canisoID'].isnull()].copy()
print(len(noxref.sUKBID.unique()))
checkColumnValues(noxref, 'isoform_group')
# all ukb stable IDs with missing xref ENSP from idat file are multi isoform IDs

# add gene name column then saving file with 56 ukb IDs that
# dont have a stable ID match or canonical ID match in idat file

noxref2 = add_genename(noxref, 'sUKBID')
noxref2.head()
#noxref2.to_csv("UKBIDs_no_idat_xref_no_canisoID_56multi_isoform_sUKBIDs.csv", index=False)

56
  isoform_group  Count
0         multi     56


sUKBID  sUKBID_xref_in_idat canisoID_xrefID_multiisoformUKBIDs  \
1    A0FGR8                False                                NaN   
18   A6NNF4                False                                NaN   
78   O14559                False                                NaN   
143  O15392                False                                NaN   
195  O43708                False                                NaN   

    isoform_group canisoID  sUKBID_xrefID HGNC_name  
1           multi      NaN            NaN     ESYT2  
18          multi      NaN            NaN    ZNF726  
78          multi      NaN            NaN  ARHGAP33  
143         multi      NaN            NaN     BIRC5  
195         multi      NaN            NaN     GSTZ1

---
---
---


**Merging files** with sukbIDs and xref ENSP and canisoIDs and xref ENSP cross-references from idat file mapping (01_2018 uniprot version)
- with idat2_stable_found, total unique sUKBID should be 3953-56

- col names of merge files : 
    - ['sUKBID', 'sUKBID_xref_in_idat', 'xref_ID', 'isoform_group',
       'canisoID']
- for single isoform IDs 

---
---
---

In [178]:
idat2_stable_found=idat2_stable_found[['sUKBID', 'sUKBID_xref_in_idat', 'sUKBID_xrefID', 'isoform_group']].copy()
# add canisoID col to idat2_stable_found- if single isoform 
    # add -0 (not actually determined) ukb fasta and idat both 
    # have no canonical isoform ID for single isoform entries
idat2_stable_found['canisoID'] = idat2_stable_found['sUKBID'].astype('str') + '-0' 
idat2_stable_found.head(3)

sUKBID  sUKBID_xref_in_idat    sUKBID_xrefID isoform_group  canisoID
0  Q04917                 True  ENSP00000248975        single  Q04917-0
1  P61981                 True  ENSP00000306330        single  P61981-0
2  P27348                 True  ENSP00000238081        single  P27348-0

In [179]:
idat2_stable_found.columns = ['sUKBID', 'sUKBID_xref_in_idat', 'sUKBID_xrefID', 'isoform_group',
       'canisoID']
checkColumnValues(idat2_stable_found, 'isoform_group')

  isoform_group  Count
0        single   2554
1         multi     62


**from stableID xref ENSP idat file, subsetted out the multi isoform ids in order to recover canonical isoform ID for these entries**
* IF the ukbID is for a multi isoform seq. i may have the canonical isoform ID
* THEN with canon iso ID, i can search for xref ENSP again in idat file using the versioned ID

In [180]:
multi_stable_found = idat2_stable_found[idat2_stable_found['isoform_group'] == 'multi']
multi_stable_found = multi_stable_found[['sUKBID','sUKBID_xref_in_idat','sUKBID_xrefID','isoform_group']].copy()
print("multi isoform group shape: ", multi_stable_found.shape)
single_stable_found = idat2_stable_found[idat2_stable_found['isoform_group'] == 'single']
print("single isoform group shape: ", single_stable_found.shape)
single_stable_found.head(3)

multi isoform group shape:  (62, 4)
single isoform group shape:  (2554, 5)


sUKBID  sUKBID_xref_in_idat    sUKBID_xrefID isoform_group  canisoID
0  Q04917                 True  ENSP00000248975        single  Q04917-0
1  P61981                 True  ENSP00000306330        single  P61981-0
2  P27348                 True  ENSP00000238081        single  P27348-0

In [181]:
# file with xref pulled from idat after 
# searching canonical isoform ID (from panel 2) ID in idat file 
# only for multi isoform uniprot IDs
sim_caniso.head(3)

sUKBID  canisoID          xref_ID
0  P31946  P31946-1  ENSP00000300161
1  P31946  P31946-1  ENSP00000361930
2  P62258  P62258-1  ENSP00000264335

In [182]:
# mapping canisoID column to multi_stable_found based on 'sUKBID'
sID_vID = sim_caniso[['sUKBID', 'canisoID']].copy()
sID_vID.drop_duplicates(inplace=True, keep='first')
idat_multi_stable = pd.merge(multi_stable_found, sID_vID, on=['sUKBID'], how='left')
print("shape before adding canisoID: ", multi_stable_found.shape)
print("shape after adding canisoID: ", idat_multi_stable.shape)
describeMe(idat_multi_stable)
idat_multi_stable

shape before adding canisoID:  (62, 4)
shape after adding canisoID:  (62, 5)
(62, 5)
Index(['sUKBID', 'sUKBID_xref_in_idat', 'sUKBID_xrefID', 'isoform_group',
       'canisoID'],
      dtype='object')
sUKBID                  0
sUKBID_xref_in_idat     0
sUKBID_xrefID           0
isoform_group           0
canisoID               26
dtype: int64


sUKBID  sUKBID_xref_in_idat    sUKBID_xrefID isoform_group  canisoID
0   Q53H12                 True  ENSP00000458417         multi  Q53H12-1
1   Q9ULX6                 True  ENSP00000380557         multi       NaN
2   Q9ULX6                 True  ENSP00000470952         multi       NaN
3   Q10567                 True  ENSP00000319361         multi       NaN
4   Q10567                 True  ENSP00000350199         multi       NaN
..     ...                  ...              ...           ...       ...
57  O15533                 True  ENSP00000372543         multi  O15533-1
58  O15533                 True  ENSP00000395701         multi  O15533-1
59  Q969X6                 True  ENSP00000458843         multi  Q969X6-1
60  Q00341                 True  ENSP00000399139         multi       NaN
61  Q07157                 True  ENSP00000484535         multi  Q07157-1

[62 rows x 5 columns]

In [183]:
# selecting rows with sUKBID xref ref but no known canisoID from panel2 analysis
noxref_multi = idat_multi_stable[idat_multi_stable['canisoID'].isnull()].copy()
print(len(noxref_multi.sUKBID.unique()))
checkColumnValues(noxref_multi, 'isoform_group')
noxref_multi.to_csv("UKBIDs_yes_idat_sUKBfound_xref_no_canisoID_26multi_isoform_sUKBIDs.csv", index=False)

13
  isoform_group  Count
0         multi     26


In [184]:
# mapping idat xref based on searching for canisoID  for multi
# isoform ukb id groups (they already have xref from stable ID search)
vID = sim_caniso[['canisoID', 'xref_ID']].copy()
vID.drop_duplicates(inplace=True, keep='first')
idat_multi_stable2 = pd.merge(idat_multi_stable, vID, on=['canisoID'], how='left')
print("shape before adding canisoID xref: ", idat_multi_stable.shape)
print("shape after adding canisoID: ", idat_multi_stable2.shape)
describeMe(idat_multi_stable2)
idat_multi_stable2
idat_multi_stable2.columns = ['sUKBID', 'sUKBID_xref_in_idat', 'sUKBID_xrefID', 'isoform_group',
       'canisoID', 'canisoID_xrefID_multiisoformUKBIDs']

shape before adding canisoID xref:  (62, 5)
shape after adding canisoID:  (164, 6)
(164, 6)
Index(['sUKBID', 'sUKBID_xref_in_idat', 'sUKBID_xrefID', 'isoform_group',
       'canisoID', 'xref_ID'],
      dtype='object')
sUKBID                  0
sUKBID_xref_in_idat     0
sUKBID_xrefID           0
isoform_group           0
canisoID               26
xref_ID                26
dtype: int64


In [185]:
single_stable_found['canisoID_xrefID_multiisoformUKBIDs'] = np.nan
print(single_stable_found.shape)
print(idat_multi_stable2.shape)
idat_stable_final = pd.concat([single_stable_found, idat_multi_stable2], axis=0)
print(idat_stable_final.shape)
# rows add up to 2718

(2554, 6)
(164, 6)
(2718, 6)


/Users/mariapalafox/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


---


In [186]:
# idat_canisoID = idat_canisoID[['sUKBID', 'sUKBID_xref_in_idat', 'xref_ID', 'isoform_group', 'canisoID']].copy()
#idat_canisoID.columns = ['sUKBID', 'sUKBID_xref_in_idat', 'canisoID_xrefID_multiisoformUKBIDs', 'isoform_group', 'canisoID']
#idat_canisoID['sUKBID_xrefID'] = np.nan
#idat_canisoID.columns
# changing order of columns
#idat_caniso_final = idat_canisoID[['sUKBID', 'sUKBID_xref_in_idat', 'sUKBID_xrefID', 'isoform_group',
      # 'canisoID', 'canisoID_xrefID_multiisoformUKBIDs']]
idat_caniso_final.head(2)

sUKBID  sUKBID_xref_in_idat  sUKBID_xrefID isoform_group  canisoID  \
0  A0AVT1                False            NaN         multi  A0AVT1-1   
1  A0FGR8                False            NaN         multi       NaN   

  canisoID_xrefID_multiisoformUKBIDs  
0                    ENSP00000313454  
1                                NaN

In [187]:
idat_stable_final.head(2)

sUKBID  sUKBID_xref_in_idat    sUKBID_xrefID isoform_group  canisoID  \
0  Q04917                 True  ENSP00000248975        single  Q04917-0   
1  P61981                 True  ENSP00000306330        single  P61981-0   

  canisoID_xrefID_multiisoformUKBIDs  
0                                NaN  
1                                NaN

In [188]:
print(idat_caniso_final.shape)
print(idat_stable_final.shape)
idat_xrefs = pd.concat([idat_caniso_final, idat_stable_final], axis=0)
print(idat_xrefs.shape)

(3523, 6)
(2718, 6)
(6241, 6)


In [138]:
2718+3523

6241

In [139]:
describeMe(idat_xrefs)

(6241, 6)
Index(['sUKBID', 'sUKBID_xref_in_idat', 'sUKBID_xrefID', 'isoform_group',
       'canisoID', 'canisoID_xrefID_multiisoformUKBIDs'],
      dtype='object')
sUKBID                                   0
sUKBID_xref_in_idat                      0
sUKBID_xrefID                         3523
isoform_group                            0
canisoID                                82
canisoID_xrefID_multiisoformUKBIDs    2636
dtype: int64


In [140]:
56+26
# 56 IDs are multi-isoform and have no canonical isoform ID and no

82

In [189]:
# map gene name to file
idat_final = add_genename(idat_xrefs, 'sUKBID')
print(idat_final.shape)
idat_final.head(3)

(6241, 7)


sUKBID  sUKBID_xref_in_idat sUKBID_xrefID isoform_group  canisoID  \
0  A0AVT1                False           NaN         multi  A0AVT1-1   
1  A0FGR8                False           NaN         multi       NaN   
2  A0JNW5                False           NaN         multi  A0JNW5-1   

  canisoID_xrefID_multiisoformUKBIDs  HGNC_name  
0                    ENSP00000313454       UBA6  
1                                NaN      ESYT2  
2                    ENSP00000279907  UHRF1BP1L

In [190]:
uniqueCount(idat_final, 'sUKBID')
#idat_final.to_csv("UKBIDs_3953_xrefENSP_from_idatfile_6241rows.csv", index=False)

len of col:  6241
len of col set:  3953



In [191]:
dropme = idat_final[(idat_final['canisoID'].isnull()) & (idat_final['sUKBID_xref_in_idat'] == False)].copy()
checkColumnValues(dropme, 'isoform_group')
uniqueCount(dropme, 'sUKBID')
#dropme
#dropme.to_csv("UKBIDs_56dropped_no_xrefENSP_from_idatfile_multiisoformUKBIDs.csv", index=False)

  isoform_group  Count
0         multi     56
len of col:  56
len of col set:  56



In [192]:
dropIDs = dropme['sUKBID'].tolist()
print(len(dropIDs))
final = addcolumnconditional(dropIDs, idat_final, 'sUKBID', 'sUKBID_56_dropped')
final.shape
uniqueCount(final, 'sUKBID')
#dropme.to_csv("UKBIDs_56dropped_no_xrefENSP_from_idatfile_multiisoformUKBIDs.csv", index=False)

56
checking labeled column values: 
  sUKBID_56_dropped  Count
0             False   6185
1              True     56

len of col:  6241
len of col set:  3953



In [193]:
almost = final[final['sUKBID_56_dropped'] == 'True']
almost.head(20)
uniqueCount(almost, 'sUKBID')
# redundancy caused by multiple gene names

len of col:  56
len of col set:  56



In [194]:
6241-56

6185

In [195]:
final2 = final[final['sUKBID_56_dropped'] == 'False']
uniqueCount(final2, 'sUKBID')

len of col:  6185
len of col set:  3897



In [196]:
3897+56

3953

In [197]:
final2.to_csv("UKBIDs_3897_idat_xrefENSP_dropped56_multiisoform_UKBIDs.csv", index=False)
final2

sUKBID  sUKBID_xref_in_idat    sUKBID_xrefID isoform_group  canisoID  \
0    A0AVT1                False              NaN         multi  A0AVT1-1   
2    A0JNW5                False              NaN         multi  A0JNW5-1   
3    A0MZ66                False              NaN         multi  A0MZ66-1   
4    A1KXE4                False              NaN         multi  A1KXE4-1   
5    A1KXE4                False              NaN         multi  A1KXE4-1   
..      ...                  ...              ...           ...       ...   
159  O15533                 True  ENSP00000395701         multi  O15533-1   
160  O15533                 True  ENSP00000395701         multi  O15533-1   
161  Q969X6                 True  ENSP00000458843         multi  Q969X6-1   
162  Q00341                 True  ENSP00000399139         multi       NaN   
163  Q07157                 True  ENSP00000484535         multi  Q07157-1   

    canisoID_xrefID_multiisoformUKBIDs  HGNC_name sUKBID_56_dropped  
0                      ENSP00000313454       UBA6             False  
2                      ENSP00000279907  UHRF1BP1L             False  
3                      ENSP00000347532      SHTN1             False  
4                      ENSP00000374565    FAM168B             False  
5                      ENSP00000387051    FAM168B             False  
..                                 ...        ...               ...  
159                    ENSP00000372685      TAPBP             False  
160                    ENSP00000402087      TAPBP             False  
161                    ENSP00000327179       UTP4             False  
162                                NaN      HDLBP             False  
163                    ENSP00000281537       TJP1             False  

[6185 rows x 8 columns]

In [206]:
# making final cross ref files to filter ensembl releases
single = final2[final2['isoform_group'] == 'single'].copy()
single_simple = single[['sUKBID', 'sUKBID_xref_in_idat', 'sUKBID_xrefID', 'isoform_group','canisoID']].copy()
describeMe(single)
print()
single_simple.dropna(inplace=True)
describeMe(single_simple)

(2563, 8)
Index(['sUKBID', 'sUKBID_xref_in_idat', 'sUKBID_xrefID', 'isoform_group',
       'canisoID', 'canisoID_xrefID_multiisoformUKBIDs', 'HGNC_name',
       'sUKBID_56_dropped'],
      dtype='object')
sUKBID                                   0
sUKBID_xref_in_idat                      0
sUKBID_xrefID                            9
isoform_group                            0
canisoID                                 0
canisoID_xrefID_multiisoformUKBIDs    2554
HGNC_name                                1
sUKBID_56_dropped                        0
dtype: int64

(2554, 5)
Index(['sUKBID', 'sUKBID_xref_in_idat', 'sUKBID_xrefID', 'isoform_group',
       'canisoID'],
      dtype='object')
sUKBID                 0
sUKBID_xref_in_idat    0
sUKBID_xrefID          0
isoform_group          0
canisoID               0
dtype: int64


In [207]:
multi = final2[final2['isoform_group'] == 'multi'].copy()
multi_simple = multi[['sUKBID', 'sUKBID_xref_in_idat', 'isoform_group','canisoID','canisoID_xrefID_multiisoformUKBIDs']].copy()
multi_simple=multi_simple[['sUKBID', 'sUKBID_xref_in_idat', 'canisoID_xrefID_multiisoformUKBIDs','isoform_group','canisoID']]
describeMe(multi)
print()
multi_simple.dropna(inplace=True)
describeMe(multi_simple)

(3622, 8)
Index(['sUKBID', 'sUKBID_xref_in_idat', 'sUKBID_xrefID', 'isoform_group',
       'canisoID', 'canisoID_xrefID_multiisoformUKBIDs', 'HGNC_name',
       'sUKBID_56_dropped'],
      dtype='object')
sUKBID                                   0
sUKBID_xref_in_idat                      0
sUKBID_xrefID                         3458
isoform_group                            0
canisoID                                26
canisoID_xrefID_multiisoformUKBIDs      26
HGNC_name                                0
sUKBID_56_dropped                        0
dtype: int64

(3596, 5)
Index(['sUKBID', 'sUKBID_xref_in_idat', 'canisoID_xrefID_multiisoformUKBIDs',
       'isoform_group', 'canisoID'],
      dtype='object')
sUKBID                                0
sUKBID_xref_in_idat                   0
canisoID_xrefID_multiisoformUKBIDs    0
isoform_group                         0
canisoID                              0
dtype: int64


In [210]:
multi_simple.head(2)

sUKBID  sUKBID_xref_in_idat canisoID_xrefID_multiisoformUKBIDs  \
0  A0AVT1                False                    ENSP00000313454   
2  A0JNW5                False                    ENSP00000279907   

  isoform_group  canisoID  
0         multi  A0AVT1-1  
2         multi  A0JNW5-1

In [211]:
single_simple.head(2)

sUKBID  sUKBID_xref_in_idat    sUKBID_xrefID isoform_group  canisoID
0  Q04917                 True  ENSP00000248975        single  Q04917-0
1  P61981                 True  ENSP00000306330        single  P61981-0

## all MULTI isoform uniprot IDs, used the canonical isoform ID to search for xref in idat file
- dropped rows if canisoID was not available from panel 2 analysis
- total 26 IDs dropped


## all SINGLE isoform uniprot IDs used the stable ID to search for xref in idat file
- if single isoform IDs were not found as stable ID in idat, they are not included in analysis 
- i think total 5 IDs lost

In [212]:
uniqueCount(single_simple, 'sUKBID')
print()
uniqueCount(multi_simple, 'sUKBID')

len of col:  2554
len of col set:  1461


len of col:  3596
len of col set:  2418



In [213]:
# compared to main text single and multi counts for 3953 IDs
lostsingle = 1466-1461
lostmulti = 2487-2418
print(lostsingle)
print(lostmulti)
print("total lost IDs", lostsingle+lostmulti)

5
69
total lost IDs 74


In [223]:
multi_simple.columns=['sUKBID', 'sUKBID_xref_in_idat', 'idat_xref',
       'isoform_group', 'canisoID']
single_simple.columns=['sUKBID', 'sUKBID_xref_in_idat', 'idat_xref',
       'isoform_group', 'canisoID']
refidat = pd.concat([single_simple, multi_simple],axis=0)
refidat

sUKBID  sUKBID_xref_in_idat        idat_xref isoform_group  canisoID
0    Q04917                 True  ENSP00000248975        single  Q04917-0
1    P61981                 True  ENSP00000306330        single  P61981-0
2    P27348                 True  ENSP00000238081        single  P27348-0
3    P27348                 True  ENSP00000371267        single  P27348-0
4    P30153                 True  ENSP00000324804        single  P30153-0
..      ...                  ...              ...           ...       ...
158  O15533                 True  ENSP00000363700         multi  O15533-1
159  O15533                 True  ENSP00000372685         multi  O15533-1
160  O15533                 True  ENSP00000402087         multi  O15533-1
161  Q969X6                 True  ENSP00000327179         multi  Q969X6-1
163  Q07157                 True  ENSP00000281537         multi  Q07157-1

[6150 rows x 5 columns]

In [224]:
uniqueCount(refidat, 'sUKBID')
uniqueCount(refidat, 'idat_xref')
uniqueCount(refidat, 'canisoID')

len of col:  6150
len of col set:  3879

len of col:  6150
len of col set:  6069

len of col:  6150
len of col set:  3879



In [225]:
reffinal = add_genename(refidat, 'sUKBID')
reffinal.to_csv("VERSION_MAPPING_UKBID_xref_ENSP_from_idatmapping_3879UKBIDs_6069ENSP.csv", index=False)

In [227]:
# multimapping of idat xref results
def setme(s):
    ls = list(s)
    finalset = set(ls)
    return finalset

# columns based on set of IDs or sequences from all releases
groupedref = reffinal.groupby('sUKBID')['idat_xref'].apply(setme)
set_columns = pd.DataFrame(groupedref)
set_columns.reset_index(inplace=True)

set_columns['count_idat_xref_unique'] = set_columns['idat_xref'].apply(lambda x: len(x))
set_columns

sUKBID                           idat_xref  count_idat_xref_unique
0     A0AVT1                   {ENSP00000313454}                       1
1     A0JNW5                   {ENSP00000279907}                       1
2     A0MZ66                   {ENSP00000347532}                       1
3     A1KXE4  {ENSP00000374565, ENSP00000387051}                       2
4     A1L0T0                   {ENSP00000263383}                       1
...      ...                                 ...                     ...
3874  Q9Y6W3                   {ENSP00000253693}                       1
3875  Q9Y6W5                   {ENSP00000483313}                       1
3876  Q9Y6W6                   {ENSP00000355866}                       1
3877  Q9Y6X8                   {ENSP00000314709}                       1
3878  Q9Y6Y8                   {ENSP00000358071}                       1

[3879 rows x 3 columns]

In [229]:
set_columns.describe()

count_idat_xref_unique
count             3879.000000
mean                 1.564578
std                  1.649978
min                  1.000000
25%                  1.000000
50%                  1.000000
75%                  2.000000
max                 35.000000

In [236]:
set_columns2 = pd.merge(set_columns, iso_dropped, how='inner', on=['sUKBID'])
set_columns2 = add_genename(set_columns2, 'sUKBID')
set_columns2=set_columns2.sort_values(['count_idat_xref_unique'],ascending=[0]).reset_index(drop=True)
set_columns2

sUKBID                                          idat_xref  \
0     O00299  {ENSP00000401292, ENSP00000406968, ENSP0000040...   
1     Q9P1U0  {ENSP00000410954, ENSP00000414720, ENSP0000041...   
2     Q9BSV6  {ENSP00000481068, ENSP00000480503, ENSP0000048...   
3     P62805  {ENSP00000489236, ENSP00000479461, ENSP0000048...   
4     O95865  {ENSP00000395372, ENSP00000403154, ENSP0000036...   
...      ...                                                ...   
3874  Q02750                                  {ENSP00000302486}   
3875  Q02790                                  {ENSP00000001008}   
3876  Q02809                                  {ENSP00000196061}   
3877  Q02833                                  {ENSP00000380713}   
3878  Q9Y6Y8                                  {ENSP00000358071}   

      count_idat_xref_unique isoform_group HGNC_name  
0                         35        single     CLIC1  
1                         32        single     ZNRD1  
2                         24        single    TSEN34  
3                         23        single      H4C1  
4                         21        single     DDAH2  
...                      ...           ...       ...  
3874                       1         multi    MAP2K1  
3875                       1        single     FKBP4  
3876                       1         multi     PLOD1  
3877                       1         multi    RASSF7  
3878                       1         multi   SEC23IP  

[3879 rows x 5 columns]

In [237]:
set_columns2.to_csv("MULTI_MAPPING_UKBID_xref_ENSP_from_idatmapping_3879UKBIDs_6069ENSP.csv", index=False)



---
---
---

In [17]:
reffinal = pd.read_csv("VERSION_MAPPING_UKBID_xref_ENSP_from_idatmapping_3879UKBIDs_6069ENSP.csv")
# filtering below file with idat ENSP cross-references from above file
combo = pd.read_csv("combo_allreleases_3953UKBIDs_diffENSPcount_52417rows_isoformgroups.csv")

In [18]:
reffinal.head(2)

sUKBID  sUKBID_xref_in_idat        idat_xref isoform_group  canisoID  \
0  Q04917                 True  ENSP00000248975        single  Q04917-0   
1  P61981                 True  ENSP00000306330        single  P61981-0   

  HGNC_name  
0     YWHAH  
1     YWHAG

In [24]:
enspfilter = reffinal[['idat_xref', 'canisoID']]
enspfilter.drop_duplicates(inplace=True)
uniqueCount(enspfilter, 'idat_xref')
uniqueCount(enspfilter, 'canisoID')

len of col:  6069
len of col set:  6069

len of col:  6069
len of col set:  3879



/Users/mariapalafox/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [23]:
test = combo.stableID_key.str.split("_", n=2, expand=True)
testpro = test[2].copy()
combo['idat_xref'] = testpro
combo.columns = ['stableID_key', 'sUKBID', 'identical_2UKBseq', 'len_ensp_minus_ukb',
       'hamming_normalized_dist', 'levenshtein_normalized_dist', 'key_8861',
       'release', 'fasta_length', 'Isoform_group', 'idat_xref']
uniqueCount(combo, 'sUKBID')
uniqueCount(combo, 'idat_xref')

len of col:  52417
len of col set:  3953

len of col:  52417
len of col set:  11946



searching for 6069 unique ENSP (mapping to 3879 ukb IDs) pulled from uniprot idat mapping file for stable (single isoform) and canonical isoform ID (multi-isoform) IDs

In [25]:
idat_filtered = pd.merge(combo, enspfilter, how='inner', on=['idat_xref'])
idat_filtered

stableID_key  sUKBID  \
0      ENSG00000004059_ENST00000000233_ENSP00000000233  P84085   
1      ENSG00000004059_ENST00000000233_ENSP00000000233  P84085   
2      ENSG00000004059_ENST00000000233_ENSP00000000233  P84085   
3      ENSG00000004059_ENST00000000233_ENSP00000000233  P84085   
4      ENSG00000004059_ENST00000000233_ENSP00000000233  P84085   
...                                                ...     ...   
29445  ENSG00000264364_ENST00000579991_ENSP00000477310  Q96FJ2   
29446  ENSG00000264364_ENST00000579991_ENSP00000477310  Q96FJ2   
29447  ENSG00000264364_ENST00000579991_ENSP00000477310  Q96FJ2   
29448  ENSG00000264364_ENST00000579991_ENSP00000477310  Q96FJ2   
29449  ENSG00000264364_ENST00000579991_ENSP00000477310  Q96FJ2   

       identical_2UKBseq  len_ensp_minus_ukb  hamming_normalized_dist  \
0                   True                   0                      0.0   
1                   True                   0                      0.0   
2                   True                   0                      0.0   
3                   True                   0                      0.0   
4                   True                   0                      0.0   
...                  ...                 ...                      ...   
29445               True                   0                      0.0   
29446               True                   0                      0.0   
29447               True                   0                      0.0   
29448               True                   0                      0.0   
29449               True                   0                      0.0   

       levenshtein_normalized_dist  key_8861  release  fasta_length  \
0                              0.0      True       85             1   
1                              0.0      True       92             1   
2                              0.0      True       94             1   
3                              0.0      True       96             1   
4                              0.0      True       97             1   
...                            ...       ...      ...           ...   
29445                          0.0      True       85             1   
29446                          0.0      True       92             1   
29447                          0.0      True       94             1   
29448                          0.0      True       96             1   
29449                          0.0      True       97             1   

      Isoform_group        idat_xref  canisoID  
0            single  ENSP00000000233  P84085-0  
1            single  ENSP00000000233  P84085-0  
2            single  ENSP00000000233  P84085-0  
3            single  ENSP00000000233  P84085-0  
4            single  ENSP00000000233  P84085-0  
...             ...              ...       ...  
29445        single  ENSP00000477310  Q96FJ2-0  
29446        single  ENSP00000477310  Q96FJ2-0  
29447        single  ENSP00000477310  Q96FJ2-0  
29448        single  ENSP00000477310  Q96FJ2-0  
29449        single  ENSP00000477310  Q96FJ2-0  

[29450 rows x 12 columns]

In [26]:
checkColumnValues(idat_filtered, 'release')
checkColumnValues(idat_filtered, 'identical_2UKBseq')

   release  Count
0       94   6066
1       96   6038
2       97   5971
3       92   5915
4       85   5460
   identical_2UKBseq  Count
0               True  29197
1              False    253


In [34]:
idat_filtered = add_genename(idat_filtered, 'sUKBID')
idat_filtered.head(2)

stableID_key  sUKBID  identical_2UKBseq  \
0  ENSG00000004059_ENST00000000233_ENSP00000000233  P84085               True   
1  ENSG00000004059_ENST00000000233_ENSP00000000233  P84085               True   

   len_ensp_minus_ukb  hamming_normalized_dist  levenshtein_normalized_dist  \
0                   0                      0.0                          0.0   
1                   0                      0.0                          0.0   

   key_8861  release  fasta_length Isoform_group        idat_xref  canisoID  \
0      True       85             1        single  ENSP00000000233  P84085-0   
1      True       92             1        single  ENSP00000000233  P84085-0   

  HGNC_name  
0      ARF5  
1      ARF5

In [35]:
idat_filtered.to_csv("idat_ENSP_filtered_comborelease_searched6069ENSP_for3879UKBID.csv", index=False)
idat_filtered_false = idat_filtered[idat_filtered['identical_2UKBseq'] == False].copy()
idat_filtered_false

stableID_key  sUKBID  \
323    ENSG00000069424_ENST00000341524_ENSP00000340824  Q13303   
1106   ENSG00000100714_ENST00000216605_ENSP00000216605  P11586   
1107   ENSG00000100714_ENST00000216605_ENSP00000216605  P11586   
1108   ENSG00000100714_ENST00000216605_ENSP00000216605  P11586   
1109   ENSG00000100714_ENST00000216605_ENSP00000216605  P11586   
...                                                ...     ...   
28629  ENSG00000175470_ENST00000455566_ENSP00000399970  Q66LE6   
28988  ENSG00000183258_ENST00000507955_ENSP00000422753  Q9UJV9   
28989  ENSG00000183258_ENST00000507955_ENSP00000422753  Q9UJV9   
29224  ENSG00000113658_ENST00000545279_ENSP00000441954  Q99717   
29229  ENSG00000113658_ENST00000545620_ENSP00000446474  Q99717   

       identical_2UKBseq  len_ensp_minus_ukb  hamming_normalized_dist  \
323                False                  15                    0.524   
1106               False                   0                    0.001   
1107               False                   0                    0.001   
1108               False                   0                    0.001   
1109               False                   0                    0.001   
...                  ...                 ...                      ...   
28629              False                 -31                    0.943   
28988              False                -310                    0.500   
28989              False                -310                    0.500   
29224              False                  -1                    0.056   
29229              False                  -1                    0.056   

       levenshtein_normalized_dist  key_8861  release  fasta_length  \
323                          0.039      True       97             5   
1106                         0.001     False       85             1   
1107                         0.001     False       92             1   
1108                         0.001     False       94             1   
1109                         0.001     False       96             1   
...                            ...       ...      ...           ...   
28629                        0.073      True       85             1   
28988                        0.498      True       96             1   
28989                        0.498      True       97             1   
29224                        0.002      True       85             1   
29229                        0.002      True       85             1   

      Isoform_group        idat_xref  canisoID HGNC_name  
323           multi  ENSP00000340824  Q13303-1    KCNAB2  
1106         single  ENSP00000216605  P11586-0    MTHFD1  
1107         single  ENSP00000216605  P11586-0    MTHFD1  
1108         single  ENSP00000216605  P11586-0    MTHFD1  
1109         single  ENSP00000216605  P11586-0    MTHFD1  
...             ...              ...       ...       ...  
28629        single  ENSP00000399970  Q66LE6-0   PPP2R2D  
28988        single  ENSP00000422753  Q9UJV9-0     DDX41  
28989        single  ENSP00000422753  Q9UJV9-0     DDX41  
29224        single  ENSP00000441954  Q99717-0     SMAD5  
29229        single  ENSP00000446474  Q99717-0     SMAD5  

[253 rows x 13 columns]

In [38]:
checkColumnValues(idat_filtered_false, 'release')
checkColumnValues(idat_filtered_false, 'Isoform_group')

uniqueCount(idat_filtered_false, 'sUKBID')
uniqueCount(idat_filtered_false, 'idat_xref')


   release  Count
0       85     63
1       97     55
2       96     51
3       94     43
4       92     41
  Isoform_group  Count
0        single    209
1         multi     44
len of col:  253
len of col set:  66

len of col:  253
len of col set:  95



In [39]:
idat_filtered_false.to_csv("idat_ENSP_filtered_comborelease_FALSEseq_identity_253row_66UKBIDs_95ENSPIDs.csv", index=False)

---
---
---

### MULTI mapping differences from using Ensembl xref mapping files (UKB to ENSP stable, release-specific) OR using UniProt idat mapping file from 01_2018 release

files:
- combo_allreleases_3953UKBIDs_diffENSPcount_52417rows_isoformgroups.csv
- idat_ENSP_filtered_comborelease_searched6069ENSP_for3879UKBID.csv

Finding the % of cross-references for 1 ENSP to 1 UKB ID using the ensembl xref files (seperate releases) and the idat file

*calculate 1:1 % of cross-references by counting non-duplicated UKB IDs in release specific or idat df*

**% 1:1 cross references**

| map file |single isoform UKBIDs |multi isoform ukb UKBIDs| unique ENSP total | unique UKB ID total|
|--|------|------------|----|--------|
|idat| 69.26% | 73.9%  | 6068 | 3879 |
|Ensembl xref 85 | 68.85%  | 19.98%    |   10183    |   3953   |
|Ensembl xref 92 |  70.05% | 13.59%    |     10395  |    3953  |
|Ensembl xref 94 |  69.27% | 12.87%   |    10612   |    3953  |
|Ensembl xref 96 |  68.21% |  12.63%  |    10663   |   3953   |
|Ensembl xref 97 |  69.17% | 12.87%   |     10564  |   3953   |

* idat file has total 2418 multi iso and 1461 single iso UKBIDs
* ensembl xref has total 1466 single and 2487 multi iso UKBIDs

code for above results below: 

In [ ]:
# os.chdir('/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/HUMAN_9606_uniprot_IDmapping/2018_08_01_IDmapping_download/idat_xref_for_ENSP_IDs/')
ensembl = pd.read_csv("combo_allreleases_3953UKBIDs_diffENSPcount_52417rows_isoformgroups.csv")
idat = pd.read_csv("MULTI_MAPPING_UKBID_xref_ENSP_from_idatmapping_3879UKBIDs_6069ENSP.csv")
print(ensembl.columns)
print(idat.columns)

In [12]:
idat.head(2)

sUKBID                                          idat_xref  \
0  O00299  {'ENSP00000401292', 'ENSP00000406968', 'ENSP00...   
1  Q9P1U0  {'ENSP00000410954', 'ENSP00000414720', 'ENSP00...   

   count_idat_xref_unique isoform_group HGNC_name  
0                      35        single     CLIC1  
1                      32        single     ZNRD1

In [22]:
test = ensembl.stableID_key.str.split("_", n=2, expand=True)
testpro = test[2].copy()
ensembl['idat_xref'] = testpro
ensembl.head(2)

stableID_key      ID  identical_2UKBseq  \
0  ENSG00000004059_ENST00000000233_ENSP00000000233  P84085               True   
1  ENSG00000004059_ENST00000000233_ENSP00000000233  P84085               True   

   len_ensp_minus_ukb  hamming_normalized_dist  levenshtein_normalized_dist  \
0                   0                      0.0                          0.0   
1                   0                      0.0                          0.0   

   key_8861  release  fasta_length Isoform_group        idat_xref  
0      True       85             1        single  ENSP00000000233  
1      True       92             1        single  ENSP00000000233

In [25]:
# multimapping for ensembl release-specific xrefs
def setme(s):
    ls = list(s)
    finalset = set(ls)
    return finalset

In [33]:
v85 = ensembl[ensembl['release'] == 85]
print("unique ukb in v85")
uniqueCount(v85, 'ID')
print("unique ensp in v85")
uniqueCount(v85, 'idat_xref')
single85 = v85[v85['Isoform_group'] == 'single']
multi85 = v85[v85['Isoform_group'] == 'multi']

# columns based on set of IDs or sequences from all releases
gsingle = single85.groupby('ID')['idat_xref'].apply(setme)
setsingle = pd.DataFrame(gsingle)
setsingle.reset_index(inplace=True)
setsingle['count_idat_xref_unique'] = setsingle['idat_xref'].apply(lambda x: len(x))
print("single isoform ID:")
checkColumnValues(setsingle, 'count_idat_xref_unique')
print()
gmulti = multi85.groupby('ID')['idat_xref'].apply(setme)
setmulti = pd.DataFrame(gmulti)
setmulti.reset_index(inplace=True)
setmulti['count_idat_xref_unique'] = setmulti['idat_xref'].apply(lambda x: len(x))
print("multi isoform ID:")
checkColumnValues(setmulti, 'count_idat_xref_unique')

print("percent single isoform ID with 1 ENSP xref : ", 1009/1466.0 *100.0)
print("percent multi isoform ID with 1 ENSP xref : ", 497/2487.0 *100.0)

unique ukb in v85
len of col:  10183
len of col set:  3953

unique ensp in v85
len of col:  10183
len of col set:  10183

single isoform ID:
    count_idat_xref_unique  Count
0                        1   1009
1                        2    276
2                        3     82
3                        4     40
4                        5     21
5                        6     10
6                        7      8
7                        8      5
8                        9      2
9                       10      2
10                      20      2
11                      21      2
12                      28      1
13                      11      1
14                      12      1
15                      14      1
16                      15      1
17                      18      1
18                      32      1

multi isoform ID:
    count_idat_xref_unique  Count
0                        2    774
1                        1    497
2                        3    494
3                       

In [34]:
v92 = ensembl[ensembl['release'] == 92]
print("unique ukb in v92")
uniqueCount(v92, 'ID')
print("unique ensp in v92")
uniqueCount(v92, 'idat_xref')
single92 = v92[v92['Isoform_group'] == 'single']
multi92 = v92[v92['Isoform_group'] == 'multi']
# columns based on set of IDs or sequences from all releases
gsingle = single92.groupby('ID')['idat_xref'].apply(setme)
setsingle = pd.DataFrame(gsingle)
setsingle.reset_index(inplace=True)
setsingle['count_idat_xref_unique'] = setsingle['idat_xref'].apply(lambda x: len(x))
print("single isoform ID:")
checkColumnValues(setsingle, 'count_idat_xref_unique')
print()
gmulti = multi92.groupby('ID')['idat_xref'].apply(setme)
setmulti = pd.DataFrame(gmulti)
setmulti.reset_index(inplace=True)
setmulti['count_idat_xref_unique'] = setmulti['idat_xref'].apply(lambda x: len(x))
print("multi isoform ID:")
checkColumnValues(setmulti, 'count_idat_xref_unique')

print("percent single isoform ID with 1 ENSP xref : ", 1027/1466.0 *100.0)
print("percent multi isoform ID with 1 ENSP xref : ", 338/2487.0 *100.0)

unique ukb in v92
len of col:  10395
len of col set:  3953

unique ensp in v92
len of col:  10395
len of col set:  10395

single isoform ID:
    count_idat_xref_unique  Count
0                        1   1027
1                        2    249
2                        3     91
3                        4     37
4                        5     21
5                        6     10
6                        7      9
7                        8      3
8                       10      3
9                       11      2
10                       9      2
11                      18      2
12                      20      2
13                      21      2
14                      32      1
15                      14      1
16                      16      1
17                      23      1
18                      24      1
19                      35      1

multi isoform ID:
    count_idat_xref_unique  Count
0                        2    885
1                        3    530
2                       

In [35]:
v94 = ensembl[ensembl['release'] == 94]
print("unique ukb in v94")
uniqueCount(v94, 'ID')
print("unique ensp in v94")
uniqueCount(v94, 'idat_xref')
single94 = v94[v94['Isoform_group'] == 'single']
multi94 = v94[v94['Isoform_group'] == 'multi']
# columns based on set of IDs or sequences from all releases
gsingle = single94.groupby('ID')['idat_xref'].apply(setme)
setsingle = pd.DataFrame(gsingle)
setsingle.reset_index(inplace=True)
setsingle['count_idat_xref_unique'] = setsingle['idat_xref'].apply(lambda x: len(x))
print("single isoform ID:")
checkColumnValues(setsingle, 'count_idat_xref_unique')
print()
gmulti = multi94.groupby('ID')['idat_xref'].apply(setme)
setmulti = pd.DataFrame(gmulti)
setmulti.reset_index(inplace=True)
setmulti['count_idat_xref_unique'] = setmulti['idat_xref'].apply(lambda x: len(x))
print("multi isoform ID:")
checkColumnValues(setmulti, 'count_idat_xref_unique')

print("percent single isoform ID with 1 ENSP xref : ", 1015/1466.0 *100.0)
print("percent multi isoform ID with 1 ENSP xref : ", 320/2487.0 *100.0)

unique ukb in v94
len of col:  10612
len of col set:  3953

unique ensp in v94
len of col:  10612
len of col set:  10612

single isoform ID:
    count_idat_xref_unique  Count
0                        1   1015
1                        2    246
2                        3     97
3                        4     44
4                        5     22
5                        6     10
6                        7      9
7                        8      5
8                       10      2
9                        9      2
10                      11      2
11                      20      2
12                      21      2
13                      32      1
14                      12      1
15                      14      1
16                      16      1
17                      18      1
18                      23      1
19                      24      1
20                      35      1

multi isoform ID:
    count_idat_xref_unique  Count
0                        2    870
1                       

In [36]:
v96 = ensembl[ensembl['release'] == 96]
print("unique ukb in v96")
uniqueCount(v96, 'ID')
print("unique ensp in v96")
uniqueCount(v96, 'idat_xref')
single96 = v96[v96['Isoform_group'] == 'single']
multi96 = v96[v96['Isoform_group'] == 'multi']
# columns based on set of IDs or sequences from all releases
gsingle = single96.groupby('ID')['idat_xref'].apply(setme)
setsingle = pd.DataFrame(gsingle)
setsingle.reset_index(inplace=True)
setsingle['count_idat_xref_unique'] = setsingle['idat_xref'].apply(lambda x: len(x))
print("single isoform ID:")
checkColumnValues(setsingle, 'count_idat_xref_unique')
print()
gmulti = multi96.groupby('ID')['idat_xref'].apply(setme)
setmulti = pd.DataFrame(gmulti)
setmulti.reset_index(inplace=True)
setmulti['count_idat_xref_unique'] = setmulti['idat_xref'].apply(lambda x: len(x))
print("multi isoform ID:")
checkColumnValues(setmulti, 'count_idat_xref_unique')

print("percent single isoform ID with 1 ENSP xref : ", 1000/1466.0 *100.0)
print("percent multi isoform ID with 1 ENSP xref : ", 314/2487.0 *100.0)

unique ukb in v96
len of col:  10663
len of col set:  3953

unique ensp in v96
len of col:  10663
len of col set:  10663

single isoform ID:
    count_idat_xref_unique  Count
0                        1   1000
1                        2    260
2                        3     95
3                        4     40
4                        5     29
5                        7     10
6                        6      9
7                        8      5
8                       10      2
9                        9      2
10                      11      2
11                      20      2
12                      21      2
13                      32      1
14                      12      1
15                      14      1
16                      16      1
17                      18      1
18                      23      1
19                      24      1
20                      35      1

multi isoform ID:
    count_idat_xref_unique  Count
0                        2    866
1                       

In [37]:
v97 = ensembl[ensembl['release'] == 97]
print("unique ukb in v97")
uniqueCount(v97, 'ID')
print("unique ensp in v97")
uniqueCount(v97, 'idat_xref')
single97 = v97[v97['Isoform_group'] == 'single']
multi97 = v97[v97['Isoform_group'] == 'multi']
# columns based on set of IDs or sequences from all releases
gsingle = single97.groupby('ID')['idat_xref'].apply(setme)
setsingle = pd.DataFrame(gsingle)
setsingle.reset_index(inplace=True)
setsingle['count_idat_xref_unique'] = setsingle['idat_xref'].apply(lambda x: len(x))
print("single isoform ID:")
checkColumnValues(setsingle, 'count_idat_xref_unique')
print()
gmulti = multi97.groupby('ID')['idat_xref'].apply(setme)
setmulti = pd.DataFrame(gmulti)
setmulti.reset_index(inplace=True)
setmulti['count_idat_xref_unique'] = setmulti['idat_xref'].apply(lambda x: len(x))
print("multi isoform ID:")
checkColumnValues(setmulti, 'count_idat_xref_unique')

print("percent single isoform ID with 1 ENSP xref : ", 1014/1466.0 *100.0)
print("percent multi isoform ID with 1 ENSP xref : ", 320/2487.0 *100.0)

unique ukb in v97
len of col:  10564
len of col set:  3953

unique ensp in v97
len of col:  10564
len of col set:  10564

single isoform ID:
    count_idat_xref_unique  Count
0                        1   1014
1                        2    250
2                        3     94
3                        4     40
4                        5     27
5                        7     10
6                        6      9
7                        8      5
8                       10      2
9                       11      2
10                      20      2
11                      21      2
12                      32      1
13                      12      1
14                       9      1
15                      14      1
16                      16      1
17                      18      1
18                      23      1
19                      24      1
20                      35      1

multi isoform ID:
    count_idat_xref_unique  Count
0                        2    873
1                       

In [19]:
print("idat unique counts")
uniqueCount(idat, 'idat_xref')
uniqueCount(idat, 'sUKBID')

print(1461+2418)
checkColumnValues(idat, 'count_idat_xref_unique')
print()

idatsingle = idat[idat['isoform_group'] == 'single']
checkColumnValues(idatsingle, 'count_idat_xref_unique')
print()
idatmulti = idat[idat['isoform_group'] == 'multi']
checkColumnValues(idatmulti, 'count_idat_xref_unique')
                  
print("percent idat with single ENSP xref : ", 2799/3879.0 *100.0)
print("percent single isoform idat with 1 ENSP xref : ", 1012/1461.0 *100.0)
print("percent multi isoform idat with 1 ENSP xref : ", 1787/2418.0 *100.0)

3879
    count_idat_xref_unique  Count
0                        1   2799
1                        2    665
2                        3    214
3                        4     85
4                        5     42
5                        6     21
6                        7     15
7                        8     11
8                       10      6
9                       11      3
10                      14      2
11                      21      2
12                       9      2
13                      20      2
14                      16      2
15                      12      1
16                      35      1
17                      18      1
18                      24      1
19                      23      1
20                      13      1
21                      19      1
22                      32      1

    count_idat_xref_unique  Count
0                        1   1012
1                        2    245
2                        3     97
3                        4     43
4       